### ISSUE TITLE
Support for arrays

### ISSUE THREAD

**Issue by GroxExMachine** [WsdlToPhp/PackageGenerator#41]
In some cases we need to setup several values of Enum type.

Original generated code does not support such way
```php
$CampaignsGetRequest->setFieldNames([
    \Campaigns\EnumType\CampaignsCampaignFieldEnum::VALUE_NAME,
    \Campaigns\EnumType\CampaignsCampaignFieldEnum::VALUE_DAILYBUDGET
]);
```

**Original**
```php
    /**
     * Set FieldNames value
     * @uses \Campaigns\EnumType\CampaignsCampaignFieldEnum::valueIsValid()
     * @uses \Campaigns\EnumType\CampaignsCampaignFieldEnum::getValidValues()
     * @param string $fieldNames
     * @return \Campaigns\StructType\CampaignsGetRequest
     */
    public function setFieldNames($fieldNames = null)
    {
        if (!\Campaigns\EnumType\CampaignsCampaignFieldEnum::valueIsValid($fieldNames)) {
            throw new \InvalidArgumentException(sprintf('Value "%s" is invalid, please use one of: %s', $fieldNames, implode(', ', \Campaigns\EnumType\CampaignsCampaignFieldEnum::getValidValues())), __LINE__);
        }
        $this->FieldNames = $fieldNames;
        return $this;
    }
```

**Modified**
```php
    /**
     * Set FieldNames value
     * @uses \Campaigns\EnumType\CampaignsCampaignFieldEnum::valueIsValid()
     * @uses \Campaigns\EnumType\CampaignsCampaignFieldEnum::getValidValues()
     * @param string $fieldNames
     * @return \Campaigns\StructType\CampaignsGetRequest
     */
    public function setFieldNames($fieldNames = null)
    {
        // ArrayOfValues
        // The existence of this block should be based on maxOccurs
        if (is_array($fieldNames)){
            foreach ($fieldNames as $fieldName) {
                if (!\Campaigns\EnumType\CampaignsCampaignFieldEnum::valueIsValid($fieldName)) {
                    throw new \InvalidArgumentException(sprintf('Value "%s" is invalid, please use one of: %s', $fieldNames, implode(', ', \Campaigns\EnumType\CampaignsCampaignFieldEnum::getValidValues())), __LINE__);
                }
            }
        } // /ArrayOfValues
        elseif (!\Campaigns\EnumType\CampaignsCampaignFieldEnum::valueIsValid($fieldNames)) {
            throw new \InvalidArgumentException(sprintf('Value "%s" is invalid, please use one of: %s', $fieldNames, implode(', ', \Campaigns\EnumType\CampaignsCampaignFieldEnum::getValidValues())), __LINE__);
        }
        $this->FieldNames = $fieldNames;
        return $this;
    }
```
This code we need to use Yandex Direct and Google AdWords.

Example
Yandex.Direct [API Campaigns](https://tech.yandex.com/direct/doc/ref-v5/campaigns/get-docpage/)
WSDL link can be found [here](https://tech.yandex.com/direct/doc/ref-v5/campaigns/campaigns-docpage/)

And what message we should have in case of error? Do "please use **one** of" still good?

**Comment by mikaelcom** [comment]
just to let you know that I don't forget your issue and that I'll be able to work on it from February 1st :wink:

**Comment by GroxExMachine** [comment]
Thank you Mikael.

As for now I use modified \WsdlToPhp\PackageGenerator\File\Struct and it is not relies on maxOccurs.

```php
    /**
     * @param PhpMethod $method
     * @param StructAttributeModel $attribute
     * @return Struct
     */
    protected function addStructMethodSetBodyForRestriction(PhpMethod $method, StructAttributeModel $attribute)
    {
        if (($model = $this->getModelFromStructAttribute($attribute)) instanceof StructModel && $model->getIsRestriction()) {
            $parameterName = lcfirst($attribute->getCleanName());
            $method
                ->addChild(sprintf('if (is_array($%s)) {', $parameterName))
                    ->addChild($method->getIndentedString(sprintf('foreach ($%s as $value) {', $parameterName), 1))

                        ->addChild($method->getIndentedString(sprintf('if (!%s::%s($value)) {', $model->getPackagedName(true), StructEnum::METHOD_VALUE_IS_VALID), 2))
                            ->addChild($method->getIndentedString(sprintf('throw new \InvalidArgumentException(sprintf(\'Value "%%s" is invalid, please use one of: %%s\', $%s, implode(\', \', %s::%s())), __LINE__);', $parameterName, $model->getPackagedName(true), StructEnum::METHOD_GET_VALID_VALUES), 3))
                        ->addChild($method->getIndentedString('}', 2))

                    ->addChild($method->getIndentedString('}', 1))
                ->addChild('}')
                ->addChild(sprintf('elseif (!%s::%s($%s)) {', $model->getPackagedName(true), StructEnum::METHOD_VALUE_IS_VALID, $parameterName))
                ->addChild($method->getIndentedString(
                    sprintf('throw new \InvalidArgumentException(sprintf(\'Value "%%s" is invalid, please use one of: %%s\', $%s, implode(\', \', %s::%s())), __LINE__);', $parameterName, $model->getPackagedName(true), StructEnum::METHOD_GET_VALID_VALUES), 1))
                ->addChild('}');
        }
        return $this;
    }
```

**Comment by GroxExMachine** [comment]
I changed name of an issue(remove "of Enum"), cause it is about arrays in general.


[Here](https://tech.yandex.com/direct/doc/ref-v5/adgroups/adgroups-docpage/) you can find WSDL link

Here the [description](https://tech.yandex.com/direct/doc/ref-v5/adgroups/add-docpage/) of an add request.

## Case 1
Take this part 
```php
{
  "method": "add",
  "params": { 
    "AdGroups": [{  /* AdGroupAddItem */
      "Name": (string), /* required */
      "CampaignId": (long), /* required */
      "RegionIds": [(long), ... ], /* required */
      "NegativeKeywords": {  /* ArrayOfString */
        "Items": [(string), ... ] /* required */
      },
      "TrackingParams": (string),
      "MobileAppAdGroup": {  /* MobileAppAdGroupAdd */
        "StoreUrl": (string), /* required */
        "TargetDeviceType": [( "DEVICE_TYPE_MOBILE" | "DEVICE_TYPE_TABLET" ), ... ], /* required */
        "TargetCarrier": ( "WI_FI_ONLY" | "WI_FI_AND_CELLULAR" ), /* required */
        "TargetOperatingSystemVersion": (string) /* required */
      },
      "DynamicTextAdGroup": {  /* DynamicTextAdGroupAdd */
        "DomainUrl": (string) /* required */
      }
    }, ... ] /* required */
  }
}
```
As you can see *AdGroups* should be an array of *AdGroupAddItem*

What do we have generated
```php
/**
 * This class stands for AddRequest StructType
 * @subpackage Structs
 */
class AddRequest extends AbstractStructBase
{
    /**
     * The AdGroups
     * Meta informations extracted from the WSDL
     * - maxOccurs: unbounded
     * - minOccurs: 1
     * @var \YandexDirect\AdGroups\StructType\AdGroupAddItem
     */
    public $AdGroups;
    /**
     * Constructor method for AddRequest
     * @uses AddRequest::setAdGroups()
     * @param \YandexDirect\AdGroups\StructType\AdGroupAddItem $adGroups
     */
    public function __construct(\YandexDirect\AdGroups\StructType\AdGroupAddItem $adGroups = null)
    {
        $this
            ->setAdGroups($adGroups);
    }
    /**
     * Set AdGroups value
     * @param \YandexDirect\AdGroups\StructType\AdGroupAddItem $adGroups
     * @return \YandexDirect\AdGroups\StructType\AddRequest
     */
    public function setAdGroups(\YandexDirect\AdGroups\StructType\AdGroupAddItem $adGroups = null)
    {
        $this->AdGroups = $adGroups;
        return $this;
    }
```
So, in this place we have no support for an array too.

## Case 2
The next case I found is this one:

Add request contain an array of long *"RegionIds": [(long), ... ], /* required */*
<xsd:element name="RegionIds" type="xsd:long" minOccurs="1" maxOccurs="unbounded"/>
This is generated set method for it.
```php
    /**
     * Set RegionIds value
     * @param long $regionIds
     * @return \YandexDirect\AdGroups\StructType\AdGroupAddItem
     */
    public function setRegionIds($regionIds = null)
    {
        $this->RegionIds = $regionIds;
        return $this;
    }
```
1. Should we write *public function setRegionIds(array $regionIds = null)* ?
2. Should we check for an array type automatically and throw exeption, something like InvalidArgumentException?

**Comment by GroxExMachine** [comment]
It is look, like *xsd:sequence* did not handled right.
```xml
 <xsd:element name="AddRequest">
            <xsd:complexType>
               <xsd:sequence>
                  <xsd:element name="AdGroups" type="ns:AdGroupAddItem" minOccurs="1" maxOccurs="unbounded"/>
               </xsd:sequence>
            </xsd:complexType>
         </xsd:element>
```
In all cases I have only *ArrayOfInteger.php* and *ArrayOfString.php* and this classes not used in any generated class. 
By WSDL, as I assume, there should be more ArraysOf____.

**Comment by mikaelcom** [comment]
Hi,

Thank you for such researches and explanations,

I just downloaded the WSDL and the schema from https://tech.yandex.com/direct/doc/ref-v5/campaigns/campaigns-docpage/ and I don't find the AddRequest element, **is it from another file ?**

In any case, you might have pointed out a case that is not well handled and that should alter a lot the generated classes regarding the sequence parsing. I must admit I might not handled it very perfectly so I'll improve (make it right) this time.

**Comment by mikaelcom** [comment]
In fact, **sequence** tags are not handled at all yet ...!

**Comment by mikaelcom** [comment]
@GroxExMachine To be sure of that has to be handled, can you confirm that only the elements that have the `maxOccurs` attribute with the value and only the value `unbounded` or any integer > 1 must be considered as a class' property that is an array? (following this thinking and the below documentation link, any element with only the `minOccurs` attribute with an integer  > 1 should be defined as an array)

For information, I found this documentation https://www.w3.org/TR/xmlschema-0/#OccurrenceConstraints that tells how the minOccurs/maxOccurs can be defined.

**Comment by GroxExMachine** [comment]
Yes, @mikaelcom, you are correct. And I have addition. 

Situation:
[Here](https://tech.yandex.com/direct/doc/ref-v5/campaigns/add-docpage/#input) _Campaigns_ is an array.

```php
{
  "method": "add",
  "params": { 
    "Campaigns": [{  /* CampaignAddItem */
```

It is described as _Anonymous Complex Type_ - _AddRequest_
```xml
<xsd:element name="AddRequest">
  <xsd:complexType>
    <xsd:sequence>
      <xsd:element name="Campaigns" type="ns:CampaignAddItem" minOccurs="1" maxOccurs="unbounded"/>
    </xsd:sequence>
  </xsd:complexType>
</xsd:element>
```
So, the _Campaigns_ is an array of _CampaignAddItem_. And this means, that an _AddRequest_ contain of only one _Campaigns_ array.
And, according to the specification, it looks like the error in Yandex Direct WSDL. But it should be investigated further.

**Comment by mikaelcom** [comment]
I agree with your last point as 
```xml
<xsd:element name="Campaigns" type="ns:CampaignAddItem" minOccurs="1" maxOccurs="unbounded"/>
```
means that `Campaigns` has to occur at least one and can occur more that once which seems odd.

**Comment by GroxExMachine** [comment]
The guys from Yandex Direct already answered me, that the WSDL is correct and there can be several _Campaigns_ in it:
...
<Campaigns>
    Campaign params
</Campaigns>
<Campaigns>
    Campaign params
</Campaigns>
<Campaigns>
    Campaign params
</Campaigns>
...

**Comment by mikaelcom** [comment]
With PHP, I don't see how to send multiple `Campaigns` objects with the _index_ `Campaigns` containing themselves one or more `CampaignAddItem` object(s) as you have only two possibilities to send the parameter with OOP*:
 - an **array** with indexes, each index must be named as the parameter name, in this case `Campaigns` which means that you can only have one entry in your array with the index `Campaigns`
 - an **object** with properties, each property must be named as the parameter name, in this case, `Campaigns` which means that you can only have on property in your object with a property named `Campaigns`

Plus, regarding the JSON structure indicated in [your link](https://tech.yandex.com/direct/doc/ref-v5/campaigns/add-docpage/#input), you also can't have multiple `Campaigns` items at the same level as either it is invalid or at least the last item will override the previous ones, am I wrong?

So after turning the problem in all the way I could in my head, the only/simpliest possible way I see is to send only one item named `Campaigns` which would be an array of `CampaignAddItem`. The proprty already exists, it only has to be enhanced.

So:
- the `Campaigns` property has to evolve to an array type of `CampaignAddItem` items (this means that a logic should be added to `setCampaigns` method in order to ensure that each item is a `CampaignAddItem` object otherwise an exception is throwned. An empty array as parameter would reset the property to an empty array, no exception would be throwned then,
- a new (utility, handy) method should be added for each property that is an array and named with the property name such as `addTo{property name}` (eg: `addToCampaigns`) with a parameter type hint matching the item value/object type (eg: `addToCampaigns(CampaignAddItem $item)`)

If I'm not wrong, having an array of `CampaignAddItem` object does send multiple `CampaignAddItem` item in the XML request (or any type of request with this same schema, as in your top first comment). Could you confirm? Plus is it well interpreted by the Web Service?

Am I missing a case/solution?

*: using pure XML, you can have multiple Campaigns elements (tag name is Campaigns) at the same level

**Comment by GroxExMachine** [comment]
> Plus, regarding the JSON structure indicated in your link, you also can't have multiple Campaigns items at the same level as either it is invalid or at least the last item will override the previous ones, am I wrong?

You are right.

> If I'm not wrong, having an array of CampaignAddItem object does send multiple CampaignAddItem item in the XML request (or any type of request with this same schema, as in your top first comment). Could you confirm?

Yes, I confirm.

> Plus is it well interpreted by the Web Service?

Setting up tests right now.

**Comment by GroxExMachine** [comment]
> the Campaigns property has to evolve to an array type of CampaignAddItem items
May be there should be ArrayOfCampaignAddItem*. Campaigns will contain object ArrayOfCampaignAddItem, and in this case we do not need to use an array as an argument and will have a better type hinting. 
The logic of classes will be cleaner and it will not differ too much from its current state.

*What do you think about plural form? ArraoyOfIntegers, ArrayOfStrings, ArrayOfCampaignAddItems?

**Comment by GroxExMachine** [comment]
> Plus is it well interpreted by the Web Service?

I checked. It is well interpreted. This request got positive response.
```json
{
  "method":"update",
  "params":{
    "Campaigns":[
      {"Id":"120699"},
      {"Id":"120700"},
      {"Id":"120701"}
    ]
  }
}
```

**Comment by mikaelcom** [comment]
@GroxExMachine good to know, thanks, 

I'll take this into account for the enhancement,

thx

**Comment by mikaelcom** [comment]
Answer/envisaged solution to your [first question](#issue-127472598)
> And what message we should have in case of error? Do "please use one of" still good?

at https://github.com/WsdlToPhp/PackageGenerator/commit/9015ef92a6b79d3e0bc808e9c08b136f7f9df027#commitcomment-15908987

**Comment by mikaelcom** [comment]
@GroxExMachine This branch should be sufficiently up to date to handle quite better than before any array case, even the previous ones that were not as good as it is now,

Can you try it by your side and tell me if you'd want something more complete? Any suggestion/feedback is appreciated,

Thx

ps: I'll certainly start a refactoring process after having published the new release with these improvements.

**Comment by GroxExMachine** [comment]
@mikaelcom adding several Campaigns by array works. Thank you!
It just small tests, that I have done. Later, there may be some additions, but for now, it is works as it should.

**Comment by GroxExMachine** [comment]
You may want, to add something. But on my side, issue can be closed.

**Comment by mikaelcom** [comment]
I think I'll review a little bit the set{property} method in order to avoid [repeating the sanity check](https://github.com/WsdlToPhp/PackageGenerator/blob/feature/issue-41/tests/resources/generated/ValidAdGroupsSelectionCriteria.php#L117) by using the [addTo method](https://github.com/WsdlToPhp/PackageGenerator/blob/feature/issue-41/tests/resources/generated/ValidAdGroupsSelectionCriteria.php#L132) that'll be the only method that'll throw an exception, in order to centralize the behaviors (except for the enumerations as we want only [one exception for multiple invalid values](https://github.com/WsdlToPhp/PackageGenerator/blob/feature/issue-41/tests/resources/generated/ValidAdGroupsSelectionCriteria.php#L200))

What do you think about that?

**Comment by mikaelcom** [comment]
After considering it, I cancelled what I said previously
> I think I'll review a little bit the set{property} method in order to avoid repeating the sanity check by ...

This is the final version now, I'll merge it soon and create a new release

**Comment by mikaelcom** [comment]
fixed in release [1.3.0](https://github.com/WsdlToPhp/PackageGenerator/releases/tag/1.3.0)

**Comment by mikaelcom** [comment]
This enhancemnt has introcued a new bug:
```php
public function setItem(array $item = array())
{
    foreach ($item as $item) {
        if (!$item instanceof \DocData\StructType\DocDataItem) {
            throw new \InvalidArgumentException(sprintf('The item property can only contain items of \DocData\StructType\DocDataItem, "%s" given', is_object($item) ? get_class($item) : gettype($item)), __LINE__);
        }
    }
    $this->item = $item;
    return $this;
}
```
